File: apps/mobile/lib/features/facility/domain/facility.dart
enum FacilityBookingStatus { confirmed, cancelled }

FacilityBookingStatus facilityBookingStatusFromApi(Object? value) {
  switch (value) {
    case 'CANCELLED':
      return FacilityBookingStatus.cancelled;
    default:
      return FacilityBookingStatus.confirmed;
  }
}

class Facility {
  const Facility({
    required this.id,
    required this.name,
    required this.openTime,
    required this.closeTime,
    required this.capacity,
    required this.rules,
  });

  final String id;
  final String name;
  final String openTime;
  final String closeTime;
  final int? capacity;
  final String? rules;

  factory Facility.fromJson(Map<String, dynamic> json) {
    return Facility(
      id: json['id'] as String,
      name: json['name'] as String,
      openTime: json['openTime'] as String,
      closeTime: json['closeTime'] as String,
      capacity: json['capacity'] as int?,
      rules: json['rules'] as String?,
    );
  }
}

class FacilityBooking {
  const FacilityBooking({
    required this.id,
    required this.facilityId,
    required this.facilityName,
    required this.bookingDate,
    required this.startTime,
    required this.endTime,
    required this.purpose,
    required this.status,
    required this.bookedByName,
    required this.houseCode,
    required this.householdDisplayName,
    required this.createdAt,
  });

  final String id;
  final String facilityId;
  final String facilityName;
  final String bookingDate;
  final String startTime;
  final String endTime;
  final String? purpose;
  final FacilityBookingStatus status;
  final String bookedByName;
  final String houseCode;
  final String householdDisplayName;
  final DateTime createdAt;

  factory FacilityBooking.fromJson(Map<String, dynamic> json) {
    return FacilityBooking(
      id: json['id'] as String,
      facilityId: json['facilityId'] as String,
      facilityName: json['facilityName'] as String,
      bookingDate: json['bookingDate'] as String,
      startTime: json['startTime'] as String,
      endTime: json['endTime'] as String,
      purpose: json['purpose'] as String?,
      status: facilityBookingStatusFromApi(json['status']),
      bookedByName: json['bookedByName'] as String,
      houseCode: json['houseCode'] as String,
      householdDisplayName: json['householdDisplayName'] as String,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }
}

File: packages/contracts/src/facility.ts
import { z } from "zod";

import { dataEnvelopeSchema } from "./envelope";

export const facilityBookingStatusSchema = z.enum(["CONFIRMED", "CANCELLED"]);

export type FacilityBookingStatus = z.infer<typeof facilityBookingStatusSchema>;

export const facilitySchema = z.object({
  id: z.string().uuid(),
  name: z.string().min(1),
  openTime: z.string().min(1),
  closeTime: z.string().min(1),
  capacity: z.number().int().nonnegative().nullable(),
  rules: z.string().nullable(),
});

export type Facility = z.infer<typeof facilitySchema>;

export const facilityListResponseSchema = dataEnvelopeSchema(
  z.object({ items: z.array(facilitySchema) }),
);

export type FacilityListResponse = z.infer<typeof facilityListResponseSchema>;

const timeSchema = z
  .string()
  .regex(/^(?:[01]\d|2[0-3]):[0-5]\d$/, "Waktu harus menggunakan format HH:mm.");

export const createFacilityBookingInputSchema = z
  .object({
    facilityId: z.string().uuid(),
    bookingDate: z
      .string()
      .regex(/^\d{4}-\d{2}-\d{2}$/, "Tanggal harus menggunakan format YYYY-MM-DD."),
    startTime: timeSchema,
    endTime: timeSchema,
    purpose: z.string().trim().max(200).optional(),
  })
  .strict()
  .refine((input) => input.startTime < input.endTime, {
    message: "Waktu selesai harus setelah waktu mulai.",
    path: ["endTime"],
  });

export type CreateFacilityBookingInput = z.infer<typeof createFacilityBookingInputSchema>;

export const facilityBookingListQuerySchema = z.object({
  facilityId: z.string().uuid().optional(),
  date: z
    .string()
    .regex(/^\d{4}-\d{2}-\d{2}$/)
    .optional(),
  limit: z.coerce.number().int().min(1).max(100).default(50),
});

export const facilityBookingSchema = z.object({
  id: z.string().uuid(),
  facilityId: z.string().uuid(),
  facilityName: z.string().min(1),
  bookingDate: z.string(),
  startTime: z.string(),
  endTime: z.string(),
  purpose: z.string().nullable(),
  status: facilityBookingStatusSchema,
  bookedByName: z.string().min(1),
  houseCode: z.string().min(1),
  householdDisplayName: z.string().min(1),
  createdAt: z.string().datetime({ offset: true }),
});

export type FacilityBooking = z.infer<typeof facilityBookingSchema>;

export const facilityBookingListResponseSchema = dataEnvelopeSchema(
  z.object({ items: z.array(facilityBookingSchema) }),
);

export const facilityBookingMutationResponseSchema = dataEnvelopeSchema(
  z.object({ booking: facilityBookingSchema }),
);

export type FacilityBookingListResponse = z.infer<typeof facilityBookingListResponseSchema>;
export type FacilityBookingMutationResponse = z.infer<typeof facilityBookingMutationResponseSchema>;
